text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,178 @@
+
+## UI
+
+For the UI, we have the following constraints:
+
+ * Must work on mobile and desktop: we cannot rely on mouse hover events.
+ * It's better if it works without javascript (it's important for accessibility, but also for maintenance reasons too: less JS code, less maintenance).
+
+Some extra explanations about the statement "less JS code, less maintenance": To have this information in the front-end, the back-end has to generate it in any case (either as a link or any other format). So deciding to use JS to handle this would mean that we still have code in the back-end but we would also have code in the front-end to treat it.
+
+With this in mind, the suggested UI would be like this:
+
+ * Links generated on idents in source code will jump to the ident definition.
+ * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
+ * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).
+
+Another thing which would be nice: to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).
+
+Basically, the source code pages' UI doesn't change.
+
+As for the size difference for the generated documentation, here are a few examples:
+
+| crate name | without the feature (in bytes) | with the feature (in bytes) | diff |
+| ---------- | ------------------------------ | --------------------------- | ---- |
+| std | 11.788.288 | 13.172.736 | 11.7% |
+| tract_core | 4.276.224 | 5.554.176 | 29.9% |
+| stm32f4 | 22.335.488 | 26.984.448 | 20.8% |
+| image | 4.927.488 | 6.045.696 | 22.7% |
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?
+
+Unfortunately, on this part, the best answer I can provide is that we will need to put clear limits to what we don't want. I think we need to keep rustdoc focused on its target: documentation. We provide source code pages, so this feature allows to browse them more efficiently by using information we already have. Anything more advanced would be very likely out-of-scope.
+
+**Another concern is**: "why should we implement this in rustdoc? Aren't there already existing tools doing it?"
+
+A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?
+
```

Comment:
another potential concern is the size of the generated code, it's worth addressing

Reply:
Good point!

Reply:
Oh actually it's already mentioned in `UI`. I really need to clean this up...